include/boost/math/statistics/detail/single_pass.hpp

```diff
@@ -0,0 +1,384 @@
+//  (C) Copyright Nick Thompson 2018
+//  (C) Copyright Matt Borland 2020
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_STATISTICS_UNIVARIATE_STATISTICS_DETAIL_SINGLE_PASS_HPP
+#define BOOST_MATH_STATISTICS_UNIVARIATE_STATISTICS_DETAIL_SINGLE_PASS_HPP
+
+#include <tuple>
+#include <iterator>
+#include <atomic>
+#include <memory>
+#include <thread>
+#include <type_traits>
+#include <future>
+#include <cmath>
+#include <algorithm>
+#include <valarray>
+#include <mutex>
+#include <unordered_map>
+
+namespace boost::math::statistics::detail
+{
+template<typename ReturnType, typename ForwardIterator>
+ReturnType mean_sequential_impl(ForwardIterator first, ForwardIterator last)
+{
+    if constexpr (std::is_same_v<typename std::iterator_traits<ForwardIterator>::iterator_category, std::random_access_iterator_tag>)
+    {
+        std::size_t elements {std::distance(first, last)};
+        std::valarray<ReturnType> mu {0, 0, 0, 0};
+        std::valarray<ReturnType> temp {0, 0, 0, 0};
+        ReturnType i = 1;
+        const auto end {last - (elements % 4)};
+
+        for(auto it {first}; it != end; it += 4)
+        {
+            const ReturnType inv {ReturnType(1) / i};
+            temp = {*it, *(it+1), *(it+2), *(it+3)};
+            temp -= mu;
+            mu += (temp *= inv);
+            i += 1;
+        }
+
+        const ReturnType num1 = ReturnType(elements - (elements %4))/ReturnType(4);
+        const ReturnType num2 = num1 + ReturnType(elements % 4);
+
+        for (auto it = end; it != last; ++it)
+        {
+            mu[3] += (*it-mu[3])/i;
+            i += 1;
+        }
+
+        return (num1 * std::valarray<ReturnType>(mu[std::slice(0,3,1)]).sum() + num2 * mu[3]) / ReturnType(elements);
+    }
+    else
+    {
+        auto it = first;
+        ReturnType mu = *it;
+        ReturnType i = 2;
+        while(++it != last)
+        {
+            mu += (*it - mu)/i;
+            i += 1;
+        }
+
+        return mu;
+    }
+}
+
+// Higham, Accuracy and Stability, equation 1.6a and 1.6b:
+// Calculates Mean, M2, and variance
+template<typename ReturnType, typename ForwardIterator>
+ReturnType variance_sequential_impl(ForwardIterator first, ForwardIterator last)
+{
+    using Real = std::tuple_element_t<0, ReturnType>;
+
+    Real M = *first;
+    Real Q = 0;
+    Real k = 2;
+    Real M2 = 0;
+
+    for(auto it = std::next(first); it != last; ++it)
+    {
+        Real tmp = (*it - M) / k;
+        Real delta_1 = *it - M;
+        Q += k*(k-1)*tmp*tmp;
+        M += tmp;
+        k += 1;
+        Real delta_2 = *it - M;
+        M2 += delta_1 * delta_2;
+    }
+
+    return std::make_tuple(M, M2, Q/(k-1));
+}
+
+// Global thread counter required for recursive calls. Reset before each parallel method is called
+static std::atomic<unsigned> thread_counter {1};
```

Comment:
I admit to not understanding why this needs a global static.

Reply:
It guarantees thread-safe initialization vs using a local static variable. It is also atomic so that each thread can safely increment it in order to avoid over-subscription since I am using `std::launch::async` policy.

Reply:
A global counter is going to cause problems. This counter can be shared by two entirely independent calculations running at the same time. Using Janine Bennet's algorithm will get you past this.

Reply:
The intent here was that it is shared by two independent calculations to avoid over-subscription. Looking back the case where a user specifies a parallel execution policy from two different threads is unlikely. I can make it locally static without any real issue. The other option would be to delete the thread counter entirely, and change the async launch policy to let the runtime deal with thread synchronization. Do you have a particular opinion either way?

Reply:
>  Looking back the case where a user specifies a parallel execution policy from two different threads is unlikely.

I don't understand why this is particularly unlikely; nor is probability a useful framing as predicting how things will be used is very difficult. But nonetheless, imagine a Rest API where the server is multithreaded and each thread uses this class to launch threads.

> The other option would be to delete the thread counter entirely, and change the async launch policy to let the runtime deal with thread synchronization. Do you have a particular opinion either way?

I admit I don't see how this is gonna work, though if you flesh it out in code I might be able to understand. Personally I think the pairwise recursive decomposition is going to cause problems-some of which you've already encountered-but I feel it's very fundamental to the nature of the algorithm, and [Bennet's](https://link.springer.com/article/10.1007/s00180-015-0637-z) algorithm will be simpler.

Reply:
I should have something up in the next day. Chan only described the calculation of M1 and M2. The current implementation of `first_four_moments` follows EQN 3.1 out of the linked paper (I had previously seen it from [Terriberry](https://web.archive.org/web/20140423031833/http://people.xiph.org/~tterribe/notes/homs.html)). Converting from recursive decomposition to linear should not be overly strenuous.